docs/02-app/01-building-your-application/01-routing/03-linking-and-navigating.mdx

```diff
@@ -183,6 +183,70 @@ For a full list of `useRouter` methods, see the [API reference](/docs/app/api-re
 
 > **Recommendation:** Use the `<Link>` component to navigate between routes unless you have a specific requirement for using `useRouter`.
 
+## Using the native History API
+
+Next.js allows you to use the native [`window.history.pushState`](https://developer.mozilla.org/en-US/docs/Web/API/History/pushState) and [`window.history.replaceState`](https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState) methods to update the browser's history stack without reloading the page.
+
+This behavior is enabled with the Next.js [`windowHistorySupport`](/docs/app/api-reference/next-config-js/windowHistorySupport) config option. When the flag is set to `true`, `pushState` and `replaceState` calls integrate into the Next.js Router, allowing you to sync with [`usePathname`](/docs/app/api-reference/functions/use-pathname) and [`useSearchParams`](/docs/app/api-reference/functions/use-search-params).
+
+### `window.history.pushState`
+
+Use it to add a new entry to the browser's history stack. The user can navigate back to the previous state. For example, to sort a list of products:
+
+```tsx fileName="app/ui/sort-products.tsx"
+'use client'
+
+import { useCallback } from 'react'
+import { useSearchParams } from 'next/navigation'
+
+export default function SortProducts() {
+  const searchParams = useSearchParams()
+
+  const updateSorting = useCallback(
+    (sortOrder: string) => {
+      const params = new URLSearchParams(searchParams.toString())
+      params.set('sort', sortOrder)
+      window.history.pushState(null, '', `?${params.toString()}`)
+    },
+    [searchParams]
+  )
+
+  return (
+    <>
+      <button onClick={() => updateSorting('asc')}>Sort Ascending</button>
+      <button onClick={() => updateSorting('desc')}>Sort Descending</button>
+    </>
+  )
+}
+```
+
+### `window.history.replaceState`
+
+Use it to replace the current entry on the browser's history stack. The user is not able to navigate back to the previous state. For example, to switch the application's locale:
+
+```tsx fileName="app/ui/locale-switcher.tsx"
+'use client'
+
+import { usePathname } from 'next/navigation'
+
+export default function LocaleSwitcher() {
```

Comment:
Suuuuuper small nits. Prefer named exports for non-pages. Prefer `function switchLocale` versus the anon function.